src/java.base/share/classes/java/util/stream/Gatherers.java

```diff
@@ -0,0 +1,596 @@
+                );
+            }
+        }
+    }
+
+    // Public built-in Gatherers and factory methods for them
+
+    /**
+     * Gathers elements into fixed-size windows. The last window may contain
+     * fewer elements than the supplied window size.
+     *
+     * <p>Example:
+     * {@snippet lang = java:
+     * // will contain: [[1, 2, 3], [4, 5, 6], [7, 8]]
+     * List<List<Integer>> windows =
+     *     Stream.of(1,2,3,4,5,6,7,8).gather(Gatherers.windowFixed(3)).toList();
+     * }
+     *
+     * @param windowSize the size of the windows
+     * @param <TR> the type of elements the returned gatherer consumes
+     *             and the contents of the windows it produces
+     * @return a new gatherer which groups elements into fixed-size windows
+     * @throws IllegalArgumentException when groupSize is less than 1
+     */
+    public static <TR> Gatherer<TR, ?, List<TR>> windowFixed(int windowSize) {
+        if (windowSize < 1)
+            throw new IllegalArgumentException("'groupSize' must be greater than zero");
+
+        return Gatherer.ofSequential(
+                // Initializer
+                () -> new ArrayList<>(windowSize),
+
+                // Integrator
+                Integrator.<ArrayList<TR>,TR,List<TR>>ofGreedy((window, e, downstream) -> {
+                    window.add(e);
+                    if (window.size() < windowSize) {
+                        return true;
+                    } else {
+                        var full = List.copyOf(window);
+                        window.clear();
+                        return downstream.push(full);
+                    }
+                }),
+
+                // Finisher
+                (window, downstream) -> {
+                    if(!downstream.isRejecting() && !window.isEmpty())
+                        downstream.push(List.copyOf(window));
+                }
+        );
+    }
+
+    /**
+     * Gathers elements into sliding windows, sliding out the most previous
+     * element and sliding in the next element for each subsequent window.
+     * If the stream is empty then no window will be produced. If the size of
+     * the stream is smaller than the window size then only one window will
+     * be emitted, containing all elements.
+     *
+     * <p>Example:
+     * {@snippet lang = java:
+     * // will contain: [[1, 2], [2, 3], [3, 4], [4, 5], [5, 6], [6, 7], [7, 8]]
+     * List<List<Integer>> windows =
+     *     Stream.of(1,2,3,4,5,6,7,8).gather(Gatherers.windowSliding(2)).toList();
+     * }
+     *
+     * @param windowSize the size of the windows
+     * @param <TR> the type of elements the returned gatherer consumes
+     *             and the contents of the windows it produces
+     * @return a new gatherer which groups elements into sliding windows
+     * @throws IllegalArgumentException when windowSize is less than 1
+     */
+    public static <TR> Gatherer<TR, ?, List<TR>> windowSliding(int windowSize) {
```

Comment:
From my experience, sliding window with windowSize = 2 is the most important and deserves the explicit support (I have such an operation in my StreamEx library and it's used quite often). Please consider adding (name subject to discussions):

`public static <T, R> Gatherer<T, ?, R> pairMap(BiFunction<T, T, R> functionToApplyToTwoAdjacentElements)`. It could be more optimal, as you only need to store the previous element. Moreover, I believe one can implement a proper combiner to make it parallel-friendly. Could be added separately after this PR is merged. I can contribute it.

Reply:
@amaembo pairMap should be straight-forward to implement. And regarding parallelization I don't think that will necessarily work since it would not be able to run incrementally (depth-first).

Reply:
Sorry, I'm not sure what do you mean by incrementally. But I've realized that to implement this you may need to push to downstream during combining. E.g., imagine the stream of numbers 1...1000, and you want to make `parallel().gather(pairMap((left, right) -> left+"->"+right)).collect(toList())`, creating a list of "1->2", "2->3", ..., "999->1000" strings. Now, let's assume that the stream was split into four parts during the parallelization: `1..250`, `251..500`, `501..750`, `751..1000`. So you create downstream accumulators and push the pairs there with integrator, like:
"1->2", ..., "249->250" into first toList() accumulator, 
"251->252", ..., "499->500" into second toList() accumulator
"501->502", ..., "749->750" into third toList() accumulator
"751->752", ..., "999->1000" into fourth toList() accumulator
At this point we have four independent gatherer states, and they keep the first and last element from the upstream. Now, you combine for example, first and second gatherer states. The first one is [first=1, last=250], and the second is [first=251, last=500]. The combined gatherer should be [first=1, last=500] and at the same time during the combining it should push the "250->251" object to the downstream, which will land into the first toList() accumulator.

Hopefully my explanation is clear. Will it be possible to extend the combiner function and provide the Downstream there as well, so combiner can also work as a finisher for the first combined chunk?

Reply:
Ah, now I see what you mean. In this specific case you might be able to create a combiner (which would also need a Downstream) such that you can add the missing combinations (the end of the "left" and the beginning of the "right"). But imagine something like `windowFixed(3)`, where the presence of 1 or 2 elements to the "left" will skew all the elements to the "right", potentially rippling through the entire stream.

I think the Preview is a perfect place to see if the combiner would be made even better with a Downstream reference or not. Logically, the output of said Downstream would be placed *after* the "left"'s output, and *before* the "right"'s output (so in the middle of the two).

But then someone might want to be able to inject elements outside of that order, but I guess that remains to be seen.

Great thoughts, @amaembo :+1:

Reply:
With windowFixed, true parallelization is not possible, unless the stream is SIZED/SUBSIZED, but this information is not available for Gatherer, so we cannot solve this at all (and this is probably a reasonable limitation of the API).

> But then someone might want to be able to inject elements outside of that order, but I guess that remains to be seen.

Not sure where one may want to inject other elements. If you want to add something after the right output, you can do this inside the finisher, or the next combiner, when the "right" will become the "left". There are no other places to insert. E.g., you cannot insert "before left", because downstream collector has no corresponding operation. To me it looks like inserting after the "left" part is the only possibility.

Reply:
@amaembo Alright, then I think we are aligned. Experimentation with combiner-downstream should be possible during the Preview-cycle